### ISSUE TITLE
Typescript support with typings of node-ssh

### ISSUE THREAD

**Issue by abdatta** [steelbrain/node-ssh#130]
Can you please add typings for this amazing module? Would be very helpful. Thanks. :)

**Comment by philjones88** [comment]
Here's some typings myself and @Yantrio came up with, feel free to improve/change but they work for us

```ts
declare module 'node-ssh' {
  namespace SSH {
    interface Config {
      // Minimal type of all the config options ssh2 accepts
      host: string;
      port: number;
      username: string;
      password?: string;
      privateKey?: string;
      onKeyboardInteractive?: () => void;
    }

    interface ConfigGiven {
      host: string;
      port?: number;
      username: string;
      password?: string;
      privateKey?: string;
      onKeyboardInteractive?: () => void | boolean;
    }

    interface PutFilesOptions {
      sftp?: object;
      sftpOptions: object;
      concurrency: number;
    }

    interface PutDirectoryOptions {
      sftp?: object;
      sftpOptions: object;
      concurrency: number;
      recursive: boolean;
      tick: (localPath: string, remotePath: string, error?: Error) => void;
      validate: (localPath: string) => boolean;
    }
  }

  class SSH {
    constructor();

    public connect(config: ConfigGiven): Promise<this>;

    public async requestShell(): Promise<object>;

    public async requestSFTP(): Promise<object>;

    public async mkdir(path: string, type: 'exec' | 'sftp' = 'sftp', givenSftp?: object = null): Promise<void>;

    public async exec(
      command: string,
      parameters: Array<string> = [],
      options: {
        cwd?: string,
        stdin?: string,
        stream?: string,
        options?: Object,
        onStdout?: (chunk: Buffer) => void,
        onStderr?: (chunk: Buffer) => void,
      } = {},
    ): Promise<string | Object>;

    public async execCommand(
      givenCommand: string,
      options: {
        cwd?: string,
        stdin?: string,
        options?: Object,
        onStdout?: (chunk: Buffer) => void,
        onStderr?: (chunk: Buffer) => void,
      } = {},
    ): Promise<{ stdout: string, stderr: string, code: number, signal?: string }>;

    public async getFile(localFile: string, remoteFile: string, givenSftp?: object = null, givenOpts?: object = null): Promise<void>;

    public async putFile(localFile: string, remoteFile: string, givenSftp?: object = null, givenOpts?: object = null): Promise<void>;

    public async putFiles(files: Array<{ local: string, remote: string }>, givenConfig: object = {}): Promise<void>;

    public async putDirectory(localDirectory: string, remoteDirectory: string, givenConfig: object = {}): Promise<boolean>;

    public dispose(): void;
  }
  export = SSH;
}
```

**Comment by GaikwadPratik** [comment]
@philjones88 ,

create PR so that @steelbrain can merge

**Comment by khusamov** [comment]
@philjones88 

Cannot find name ConfigGiven

error TS1040: 'async' modifier cannot be used in an ambient context.

and:

![image](https://user-images.githubusercontent.com/4146998/65608941-a1ea3200-dfb7-11e9-939d-6c0e8c04e9d7.png)

**Comment by steelbrain** [comment]
I'll add these when next free, could be in a few weeks (unless I get a PR :) )

**Comment by steelbrain** [comment]
Whoops, this slipped by, taking care of this now

**Comment by steelbrain** [comment]
Rewritten in Typescript (with new features like `withShell` / `withSFTP` and `getDirectory`) in https://github.com/steelbrain/node-ssh/pull/297

**Comment by steelbrain** [comment]
Released in v9.0.0
